### ISSUE TITLE
OcRunner  needs to handle spaces in quoted arguments

### ISSUE THREAD

**Issue by NickLaMuro** [minishift/minishift#729]
Given the following addon:

**./tmp/minishift_addons/foo/foo.adon**
```bash
# Name: foo
# Description: create foo project

oc adm new-project foo --description='Bar Baz'
```

You would get the following error when attempting to boot minishift with that addon enabled:

```console
$ minishift addons install ./tmp/minishift_addons/foo
$ minishift addons enable foo
$ minishift start
...
Applying addon foo:
-- adm new-project foo --description='Bar Baz'
error: you must specify one argument: project name
See 'oc adm new-project -h' for help and examples.
E0407 17:09:06.185670   20557 start.go:232] Error executing addon commands Error executing command oc adm new-project foo --description='Bar Baz'.
```

Bug Location
-------------
I think the problem is within the `strings.Split()` that happens in to locations in the `pkg/minishift/oc/oc.go` file:

- https://github.com/minishift/minishift/blob/cc4c2a2/pkg/minishift/oc/oc.go#L56
- https://github.com/minishift/minishift/blob/cc4c2a2/pkg/minishift/oc/oc.go#L65

The argument of `oc adm new-project foo --description='Bar Baz'` would be split into something like this:

```go
["oc", "adm", "new-project", "foo", "--description='Bar", "Baz'"]
```

Which, when passed to `os/exec`'s `Run()` function, would wrap all args in quotations and attempt to run it.  When `oc` parses the command, it sees both `"foo"` and `"Baz'"` as inputs for the project name, and returns that error accordingly (`"--description='Bar"` would also probably be considered malformed, but hard to say based on the output alone).

Possible solution
-----------------
I have put together a lib that _**might**_ be a better alternative to the `strings.Split()` here:

https://github.com/NickLaMuro/go_cmd_arg_parser

As the README in the project suggests, I am neither an experienced programmer in `go`, nor the original contributor to that function (shamelessly stolen from stackoverflow; again see README), so would advise at the very least vetting this code heavily before considering implementing it, if not re-writing it yourself.

That said, it also handles the corner case of passing in escaped quotes into a string, which is something I expect a user might do.

**Comment by hferentschik** [comment]
@NickLaMuro, thanks for the bug report and this very thorough digging. You are right, this is an issue with the _oc_ runner helper. So in fact, this will not only affect add-ons, but also any other place in the code where we use the runner. We need to handle this better indeed. We will look at your suggested approach.

**Comment by hferentschik** [comment]
Looking at the code, I find it quite hard to take in. It might be too simplistic, but could one not just split the string as we do now, them iterate over it and if a token starts with a quote, but has no end quote, you join it with the next token (recursively until you find the token with the closing quote or you run out of tokens).

**Comment by NickLaMuro** [comment]
> Looking at the code, I find it quite hard to take in.

@hferentschik yeah, I admit that it took me some time to grok the code that this was originally inspired from, and the modulus division that I added made this quite a bit more complex and confusing, so I understand the confusion.  I will walk through it briefly, to hopefully provide some clarity:

1. [The first part of the function](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L13-L15) just sets up the default values for the method, including the anonymous function, `splitWithQuotes` to be passed into `strings.FieldsFunc`, which determines when to split a string.  
    * The `lastQuote` is the most recently matched quote `rune` (if we match a `'`, we want to line it up to another `'`, not a `"`).
    * The `consecutiveEscapeCount` is the number of escape runes in a row that have been a `\`.  Reset to zero at each step unless the current character is a `\`, then it is incremented
2. `splitWithQuotes` basically is prioritized list of conditions that determine when we split.  Specifically (in order):
    - [[ref]](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L17-L22) Don't split if the character matches the "last matched quote", but, reset the `lastQuote` var that holds the matched quote (could be `'`, `"`, or others), to a `null` byte (`rune(0)`) **IF** their isn't an escaped sequence that cancels it out (`\\"` would reset the token as the escape chars prior to it aren't affecting the quote, but `\"` and `\\\"` would not).
    - [[ref]](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L23-L25) Don't split if we found a `\`, but increment the counter for it (`consecutiveEscapeCount`).
    - [[ref]](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L26-L28) Don't split if `lastQuote` current value is not `rune(0)` (in other words, we are still inside a string
    - [[ref]](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L29-L32) Don't split if the character is a quote type rune, and set the `lastQuote` value to that
    - [[ref]](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L33-L35) **Do** split the string if it is a space, and any of the above haven't been met.  Reset `consecutiveEscapeCount` value here as well (`lastQuote` would have already been reset at this point).
3. Trim each of the args in the array from having outer quotes, if there are any, using the [`trimOuterQuotesFromArgs`](https://github.com/NickLaMuro/go_cmd_arg_parser/blob/master/go_cmd_arg_parser.go#L39) method.  This basically loops through each arg, determines if that arg starts and ends with the same quote, and change that arg to the substring of it with the first and last char removed.

> It might be too simplistic, but could one not just split the string as we do now, them iterate over it and if a token starts with a quote, but has no end quote, you join it with the next token (recursively until you find the token with the closing quote or you run out of tokens).

@hferentschik I think then you run into the problem of when your cmd flag includes the string argument along with it, something like this:

```
oc new-project foo --description="something foo"
```

Since `--description="something foo"` doesn't start with a string, the logic you described would fail, and you would get to args of `--description="something` and `foo"`.  Aside from wanting to avoid recursion, the logic that would then be needed to solve for this edge case using your algorithm above would increase the complexity a bit.  The algorithm I have would handle this case properly (though, I probably should add a test for this case...), but since the strings wouldn't be stripped in this case (probably the correct action though), I don't know how that would do when passed into `os/exec#Run()`.

That said, not trying to paint the picture that my proposal is perfect (did I say that I pulled it from stackoverflow... 😅 ), and so feel free to use the tests to inject your own algorithm to find something that works better.  The tests are written in a way that they just pass in an input, and expect a specific output, so you should be able to inject whatever algorithm you please into `ParseFromCmdString` and see if it fits the spec defined in the tests.

**Comment by hferentschik** [comment]
@NickLaMuro, thanks for the pointers. My idea was really a bit too simple. I also understand your code better now. We will add something like this, however, I don't think I'd like to include this as a dependency. Instead I would just inline this function and test into the Minishift code directly. I hope this  is ok with you.

**Comment by NickLaMuro** [comment]
@hferentschik No problem, in fact, that was the intent.  I figured creating a PR was overstepping a bit to much based on the contributing guidelines (and I didn't really know the best spot for it and some tests surrounding it), but figured a code example might be easier to explain some of the nuances of the problem.  Feel free to use as you see fit (assuming it doesn't both the original poster... though, by putting it on StackOverflow, you kind of give up your rights to say no as far as I can tell).

**Comment by jorgemoralespou** [comment]
@NickLaMuro To be honest, if this is good as @hferentschik says you should
do a PR. You deserve the credit.

**Comment by praveenkumar** [comment]
Fix by #764
